infrastructure/devicepathresolver/iscsi_device_path_resolver.go

```diff
@@ -0,0 +1,184 @@
+package devicepathresolver
+
+import (
+	"fmt"
+	"path"
+	"path/filepath"
+	"regexp"
+	"strings"
+	"time"
+
+	boshopeniscsi "github.com/cloudfoundry/bosh-agent/platform/openiscsi"
+	boshsettings "github.com/cloudfoundry/bosh-agent/settings"
+	boshdirs "github.com/cloudfoundry/bosh-agent/settings/directories"
+	bosherr "github.com/cloudfoundry/bosh-utils/errors"
+	boshlog "github.com/cloudfoundry/bosh-utils/logger"
+	boshsys "github.com/cloudfoundry/bosh-utils/system"
+)
+
+// iscsiDevicePathResolver resolves device path by performing Open-iscsi discovery
+type iscsiDevicePathResolver struct {
+	diskWaitTimeout time.Duration
+	runner          boshsys.CmdRunner
+	openiscsi       boshopeniscsi.OpenIscsi
+	fs              boshsys.FileSystem
+	dirProvider     boshdirs.Provider
+	logTag          string
+	logger          boshlog.Logger
+}
+
+func NewIscsiDevicePathResolver(
+	diskWaitTimeout time.Duration,
+	runner boshsys.CmdRunner,
+	openiscsi boshopeniscsi.OpenIscsi,
+	fs boshsys.FileSystem,
+	dirProvider boshdirs.Provider,
+	logger boshlog.Logger,
+) DevicePathResolver {
+	return iscsiDevicePathResolver{
+		diskWaitTimeout: diskWaitTimeout,
+		runner:          runner,
+		openiscsi:       openiscsi,
+		fs:              fs,
+		dirProvider:     dirProvider,
+		logTag:          "iscsiResolver",
+		logger:          logger,
+	}
+}
+
+func (ispr iscsiDevicePathResolver) GetRealDevicePath(diskSettings boshsettings.DiskSettings) (string, bool, error) {
+	var lastDiskID string
+
+	if diskSettings.InitiatorName == "" {
+		return "", false, bosherr.Errorf("iSCSI InitiatorName is not set")
+	}
+
+	if diskSettings.Username == "" {
+		return "", false, bosherr.Errorf("iSCSI Username is not set")
+	}
+
+	if diskSettings.Password == "" {
+		return "", false, bosherr.Errorf("iSCSI Password is not set")
+	}
+
+	if diskSettings.Target == "" {
+		return "", false, bosherr.Errorf("iSCSI Iface Ipaddress is not set")
+	}
+
+	existingPaths := []string{}
+
+	result, _, _, err := ispr.runner.RunCommand("dmsetup", "ls")
+	if err != nil {
+		return "", false, bosherr.WrapError(err, "Could not determining device mapper entries")
+	}
+
+	lastDiskID, err = ispr.lastMountedCid()
+	if err != nil {
+		return "", false, bosherr.WrapError(err, "Fetching last mounted disk CID")
+	}
+
+	ispr.logger.Debug(ispr.logTag, "Last mounted disk CID: '%s'", lastDiskID)
+
+	if !strings.Contains(result, "No devices found") {
```

Comment:
W.r.t. complexity comment above. This statement is nested three levels deep and with a for.

Any way to simplify or break it down into smaller functions?